### DISCUSSION TITLE
Plugin support for Vanilla Extract

### DISCUSSION THREAD

**Discussion by mattcompiles** [vercel/next.js#49892] [Help] [answered]
### Summary

Hi 👋 

I work on [Vanilla Extract](https://vanilla-extract.style/) (VE), a build time CSS-in-JS solution. Our adoption has been growing quite well lately, I noticed we even get a [shout out in the Next docs](https://nextjs.org/docs/app/building-your-application/styling/css-in-js) now 😄. I'm making this post to reach out for help from the Next team. 

As VE integrates into the bundler, we support Next via a "plugin" which just updates the webpack config ([source code](https://github.com/vanilla-extract-css/vanilla-extract/blob/master/packages/next-plugin/src/index.ts)). Adding the VE Webpack plugin is easy enough, however our main difficulty is handling the CSS that VE outputs. For each VE file in a project, there will be a `.vanilla.css` import injected at build time. This is just a standard CSS file however you can think of it similar to a CSS module as the class names are scoped, this is already done by VE before the bundler sees the file. We can't just drop the import and leave the rest to Next though, as next has rules around how CSS is handled. To make this work we have had to reach into [Next private internals](https://github.com/vanilla-extract-css/vanilla-extract/blob/c842fde4b1d96da8744b2993e378e23ea4e3e4cd/packages/next-plugin/src/index.ts#LL4C17-L4C17). Obviously this is very unideal, and leads to a lot of issues being opened against VE (exampoles linked below). The issues have gotten worse of late with the introduction of the app directory and these private APIs needing knowledge whether that is in use for the project or not. 

While I'm not expecting Next to support VE out of the box, I'd really like some help from the Next team to allow CSS tools like VE to support Next without doing massive hacks to integrate. 

Next specific issues:
- [Next.js 13 - \`experimental.appDir\` still being used in stable App Router](https://github.com/vanilla-extract-css/vanilla-extract/issues/1087)
- [Shared CSS is bundled multiple times with App Router in Next](https://github.com/vanilla-extract-css/vanilla-extract/issues/1088)
- [Sprinkles not applying in production build with NextJS App Directory](https://github.com/vanilla-extract-css/vanilla-extract/issues/1085)

**Comment by SuttonJack** [comment, marked as answer]
## Additional Context

> `Oct, 2022`
- Next.js 13 release with app directory

> `Feb, 2023`

-  @shuding provides some insight [here](https://github.com/vanilla-extract-css/vanilla-extract/issues/1013#issuecomment-1446302154) on updating our plugin to read the new `hasAppDir` field

> `April, 2023`

- @shuding [Updates](https://github.com/vercel/next.js/pull/48306) the CSS loader within next to provide more flexibility for libraries like VE
- We update the plugin [here](https://github.com/vanilla-extract-css/vanilla-extract/pull/1060) to introduce app directory support

> `May 9, 2023`

- `vanilla-extract` is added to the CSS-in-JS section of the [docs](https://nextjs.org/docs/app/building-your-application/styling/css-in-js) 

<h1></h1>

## Questions for the Next team

- The VE plugin [imports](https://github.com/vanilla-extract-css/vanilla-extract/blob/master/packages/next-plugin/src/index.ts) `getGlobalCSS` loader from next internals at the moment. Is there a more stable alternative here? Can we add `|css\.ts` to the regexCSS [here](https://github.com/vercel/next.js/blob/canary/packages/next/src/build/webpack/loaders/utils.ts#L20)?
    
  - Note that the current [VE demo repository](https://github.com/SuttonJack/vanilla-extract-app-dir) only works if `appDir: enabled` is specified in `next.config.js`

**Reply by mattcompiles** [reply]
Hey @leerob, why did you mark this as answered? This post is just adding context to the original question. We'd really love a response from someone on the Next team.

**Comment by exneval** [comment]
Hello, @shuding. Kindly request your attention to this. Would you please take a moment to look into it? Thank you.

**Comment by shuding** [comment]
For these three issues,

> Next.js 13 - experimental.appDir still being used in stable App Router

I opened https://github.com/vanilla-extract-css/vanilla-extract/pull/1099 to address that one!

> Shared CSS is bundled multiple times with App Router in Next

Wrote a reply [here](https://github.com/vanilla-extract-css/vanilla-extract/issues/1088#issuecomment-1563931144), and will get back to you once we have any progress. Feel free to open is as a Next.js issue and close it or redirect it here.

Update: https://github.com/vercel/next.js/pull/50406 should fix it

> Sprinkles not applying in production build with NextJS App Directory

We have the issue opened [here](https://github.com/vercel/next.js/issues/49817) as well, need to investigate still but it's being tracked.

Thanks for letting us know!

**Comment by dannyrb** [comment]
> To make this work we have had to reach into [Next private internals](https://github.com/vanilla-extract-css/vanilla-extract/blob/c842fde4b1d96da8744b2993e378e23ea4e3e4cd/packages/next-plugin/src/index.ts#LL4C17-L4C17).

@mattcompiles / @SuttonJack, is there specific work that needs to be done to address this need? This particular dependency has impacted us, and I'd like to help if at all possible. We see this build error when using `yarn@4` in `pnp` mode:

```
my-project:build: > Build error occurred
my-project:build: Error: @vanilla-extract/next-plugin tried to access next (a peer dependency) but it isn't provided by its ancestors; this makes the require call ambiguous and unsound.
my-project:build:
my-project:build: Required package: next (via "next/dist/build/webpack/config/blocks/css")
```

**Reply by dannyrb** [reply]
It looks like our issue was specific to the following scenario:

- An internal library that depends on `@vanilla-extract/next-plugin`
- This internal library does not specify `next` as a `peerDependency`
- The internal library is consumed in a monorepo by multiple projects

This caused a handful of issues that were ultimately resolved by cleaning up our internal library's peerDependencies 😅 

Cheers!
